### DISCUSSION TITLE
Floating-point atomics

### DISCUSSION THREAD

**Discussion by vrpascuzzi** [AdaptiveCpp/AdaptiveCpp#521] [Q&A] [answered]
Is `fetch_add` supported for `float` on GPUs? I've tried parsing the documentation to get a clear answer with no luck. For example,
```
#define HIPSYCL_EXT_FP_ATOMICS
#include <CL/sycl.hpp>
<...>
  // Allocate memory for an int on host and device
  std::vector<float> host_data(1);

  // Initialize host int
  host_data[0] = 1.1;

  // Create buffer from host data
  cl::sycl::buffer<float, 1> buf_host(host_data.data(), cl::sycl::range<1>(1));

  // Modify the data on-device
  auto ev = queue.submit([&](cl::sycl::handler& h) {
    auto acc = buf_host.template get_access<cl::sycl::access::mode::atomic>(h);
    h.single_task([=]() { acc[0].fetch_add(3.5f); });
  });
  ev.wait();
<...>
```
This compiles fine for the device pass but fails compiling for host:
```
$ syclcc -O2 -Wno-ignored-attributes -DSYCL_TARGET_AMD --hipsycl-targets=hip:gfx900,gfx906,gfx908 -o test_atomic test_atomic.cc
In file included from test_atomic.cc:4:
In file included from /bld4/opt/hipSYCL/bin/../include/CL/sycl.hpp:31:
In file included from /bld4/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/sycl.hpp:47:
In file included from /bld4/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/libkernel/accessor.hpp:54:
/bld4/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/libkernel/atomic.hpp:119:12: error: address argument to atomic operation must be a pointer to integer or pointer ('float *' invalid)
    return __atomic_fetch_add(_ptr, operand, __ATOMIC_RELAXED);
           ^                  ~~~~
test_atomic.cc:116:34: note: in instantiation of function template specialization 'hipsycl::sycl::atomic<float, hipsycl::sycl::access::address_space::global_space>::fetch_add<float, nullptr>' requested here
    h.single_task([=]() { acc[0].fetch_add(3.5f); });
                                 ^
1 error generated when compiling for host.
```
If I wrap the submit in `#ifdef SYCL_DEVICE_ONLY` both passes succeed, but the variable is not updated. Am I missing something trivial here? An example would be greatly appreciated.

**Comment by illuhad** [comment, marked as answer]
For the CPU backend, clang support of the `__atomic_fetch_add` builtin for floating point types is still being worked on: https://reviews.llvm.org/D71726
Once this is in, it should work on CPU. I don't know what the state is in gcc or other OpenMP compilers that you could use the hipSYCL CPU backend with.

I'm not sure why the variable is not updated on GPU. IMO it should work. If you define `HIPSYCL_EXT_FP_ATOMICS`, it just calls HIP/CUDA `atomicAdd()` in the device pass which should support floating point types. Which device are you testing with?
Can you try a regular read-write accessor and then constructing the `sycl::atomic` on top of a pointer from that accessor?

A rework of atomic support in hipSYCL is quite high on the list of my todos, as we need some improvements here for oneDPL and also should move to SYCL 2020 `atomic_ref`.

EDIT: Did you really wrap the `submit()` in `SYCL_DEVICE_ONLY`? Like
```c++
#ifdef SYCL_DEVICE_ONLY
q.submit(...);
#endif
```
If so, it can't work because you are removing the kernel invocation on the host side, so the kernel would never be called. You need to `#ifdef` the content of the `parallel_for` lambda.

**Reply by vrpascuzzi** [reply]
Thanks for the reply.

> I'm not sure why the variable is not updated on GPU. IMO it should work. If you define `HIPSYCL_EXT_FP_ATOMICS`, it just calls HIP/CUDA `atomicAdd()` in the device pass which should support floating point types. Which device are you testing with?

Good -- I am understanding correctly then. The device is an MSI Vega 56 GPU, and I am using hipSYCL 0.9.0 installed from RPMs. Integral atomics work fine, i.e. change `float` to `int` in the code above.

> Can you try a regular read-write accessor and then constructing the `sycl::atomic` on top of a pointer from that accessor?

Compiling with `-std=c++20`:
```
test_atomic.cc:113:9: error: no member named 'fetch_add' in 'std::atomic<float>'
    ptr.fetch_add(3.5f, std::memory_order_relaxed); 
    ~~~ ^
1 error generated when compiling for gfx900.
```

> You need to `#ifdef` the content of the `parallel_for` lambda.

I had tried this as well:
```
Memory access fault by GPU node-2 (Agent handle: 0x22c3ef0) on address 0x21f557000. Reason: Page not present or supervisor privilege.
Aborted (core dumped)
```

**Reply by illuhad** [reply]
> test_atomic.cc:113:9: error: no member named 'fetch_add' in 'std::atomic<float>'
    ptr.fetch_add(3.5f, std::memory_order_relaxed); 
    ~~~ ^
1 error generated when compiling for gfx900.

It seems you are using `std::atomic`, not `sycl::atomic`? You don't need C++20 for this.

> I had tried this as well:
>
> Memory access fault by GPU node-2 (Agent handle: 0x22c3ef0) on address 0x21f557000. Reason: Page not present or supervisor  privilege.
> Aborted (core dumped)

Probably what has happend is that the accessor is no longer used inside the kernel on the host side, therefore it is no longer captured, therefore your kernel arguments that are passed in on the host side don't match the device side, and then terrible things happen.
(Note that this is not hipSYCL-specific, but a general issue of how C++ lambdas work - other SYCL implementations such as DPC++ would have exactly the same issues).

Make sure to only `#ifdef` no more than necessary (you might have to introduce some dummy use of the accessor), or capture the accessors explicitly: `h.single_task([acc](){...});`

**Reply by vrpascuzzi** [reply]
> It seems you are using `std::atomic`, not `sycl::atomic`? You don't need C++20 for this.

:facepalm: 

> Probably what has happend is that the accessor is no longer used inside the kernel on the host side, therefore it is no longer captured, therefore your kernel arguments that are passed in on the host side don't match the device side, and then terrible things happen.

Good call. This makes sense.

> capture the accessors explicitly: `h.single_task([acc](){...});`

This was the key -- thanks a lot!
